### ISSUE TITLE
Improvements to log drain (host metrics)

### ISSUE THREAD

**Issue by jeffblake** [DataDog/heroku-buildpack-datadog#188]
While I understand the Heroku Log drain is separate from the buildpack, this seems to be the only public forum for Heroku users of Datadog, so I thought I'd try my luck here and hopefully rally some support!

The log drain is a good start:
https://docs.datadoghq.com/logs/guide/collect-heroku-logs/

But it has some holes:

1. Logs are not correlated to the host or service. This is a pretty big one. Right now we have to specify the service and host as static variables when adding the drain, however both of those attributes are dynamic in reality. The ability to correlate metrics, logs, errors, APM, and infrastructure together is Datadog's biggest selling point, and it's not possible for Heroku users. The infrastructure host map is also rendered useless due to this limitation. The `dyno` tag workaround suggestion is a hack
Proposal: The Heroku drain should automagically work out of the box and correlate each log line to it's correct host and service, based on using e.g `web.1`, `worker.1` as host names. The data is there to do it.

2. The runtime metric logs parser is using the incorrect `dyno` attribute. It should in fact be the `source`. e.g. for router logs, the `dyno` attribute is `web.1` (good), but for runtime metrics the `dyno` attribute is `heroku.****-*****....` (bad), the `source` attribute should be used instead. This should be a simple log parser fix change.
See comment here: https://github.com/DataDog/heroku-buildpack-datadog/issues/35#issuecomment-467244565

3. Heroku users like things easy :) Those dyno runtime metrics should be automatically linked to the Datadog infrastructure, i.e. the host metrics (cpu, etc). (and `DD_DISABLE_HOST_METRICS ` should honestly default to true, no one wants that).

4. Out of the box "Heroku Dashboard", similar to how they exist for Sidekiq, etc, which pulls in the router, and runtime metrics logs from all dynotype's, and database runtime metrics as well such as Postgres

5. The Heroku log parser does not have a rule for parsing Postgres or Redis runtime metrics. https://devcenter.heroku.com/articles/heroku-postgres-metrics-logs

Adding this level of polish to the Heroku integration would be amazing.

Thanks! 

https://devcenter.heroku.com/articles/log-runtime-metrics

**Comment by arapulido** [comment]
@jeffblake Thanks a lot for your suggestions! As you said, the log drain is not part of the buildpack. I will pass this to the logs team, but I also suggest you to open a support ticket about it. Thanks!

**Comment by jeffblake** [comment]
If anyone is interested, I opened a feature request ticket with Datadog. 

If you'd like to support it, feel free to contact Datadog and reference ticket #358557. I fear that without strong support, it will get lost in the void.

Summary of request:

- The drain should automatically correlate each log to it's correct host and service
- The runtime metric logs parser should use the source attribute
- Runtime metrics should be linked to host metrics in Datadog
- DD_DISABLE_HOST_METRICS should be true by default
- Out of the box Heroku Dashboard, functionally similar to Sidkiq's, 
- A Heroku log parser rule for parsing Postgres or Redis runtime metrics.

Note from DD:
To help me bolster your points, would you mind sending over a brief explanation of your use case- specifically what type of applications you are hosting on Heroku and why they require a more enhanced, unified, integration? 

@arapulido Thanks for accepting this request on your issue tracker, if you don't mind to leave it up for a little while for visibility, I'd appreciate it =)

**Comment by carltondickson** [comment]
Have also raised a support ticket for this one.
We're missing a good few metrics and the 2 x default dashboards don't seem to work well without them.

**Comment by jeffblake** [comment]
Yeah I've been stuck on this since. Infrastructure section is useless, and can't correlate logs/traces to host metrics. What is needed is a host remapper to map the syslog.procid (web.1, worker.1) to the host (app-name.web.1, app-name.worker.1). I've already generated metrics from logs and mapped the cpu & memory runtime metrics to the appropriate system metric (system.load.1, etc) so once I'm able to map the host, the infrastructure page should magically start working!

**Comment by edeesis** [comment]
We also have issues with multi-line logs not being parsed correctly. Heroku has a limit of 10000 bytes, then it splits the log lines: https://devcenter.heroku.com/articles/logging#log-format

For JSON logs, Datadog can't properly parse the logs, so log event attributes are missing.

We can't do anything to update the Datadog Agent like described here: https://www.datadoghq.com/blog/multiline-logging-guide/#datadog because of it being a log drain.

**Comment by arapulido** [comment]
> We also have issues with multi-line logs not being parsed correctly. Heroku has a limit of 10000 bytes, then it splits the log lines: https://devcenter.heroku.com/articles/logging#log-format
> 
> For JSON logs, Datadog can't properly parse the logs, so log event attributes are missing.
> 
> We can't do anything to update the Datadog Agent like described here: https://www.datadoghq.com/blog/multiline-logging-guide/#datadog because of it being a log drain.

@edeesis You can also use the Datadog agent to send logs to Datadog. You can read about the differences when sending through the log drain and the agent here: https://docs.datadoghq.com/agent/guide/heroku-ruby/#logs

**Comment by edeesis** [comment]
> @edeesis You can also use the Datadog agent to send logs to Datadog. You can read about the differences when sending through the log drain and the agent here: [docs.datadoghq.com/agent/guide/heroku-ruby/#logs](https://docs.datadoghq.com/agent/guide/heroku-ruby/#logs)

Appreciate the quick response. I'll give it a try. My primary concern was I wanted to keep the log drain and was worried about duplicated logs.

**Comment by arapulido** [comment]
> > @edeesis You can also use the Datadog agent to send logs to Datadog. You can read about the differences when sending through the log drain and the agent here: [docs.datadoghq.com/agent/guide/heroku-ruby/#logs](https://docs.datadoghq.com/agent/guide/heroku-ruby/#logs)
> 
> Appreciate the quick response. I'll give it a try. My primary concern was I wanted to keep the log drain and was worried about duplicated logs.

To avoid duplicated logs, you can always filter out the logs coming from the log drain that are not Heroku platform specific (which are the only ones that the Datadog agent cannot send). The live view will contain them all, but the archive will only keep the ones you haven't filtered out: https://docs.datadoghq.com/logs/guide/getting-started-lwl/#3-create-a-log-pattern-exclusion-filter

**Comment by johngallagher** [comment]
I've asked Heroku about this and the way of sending logs via the agent in https://docs.datadoghq.com/agent/guide/heroku-ruby/#logs is not supported by them.

Their response:

> Is this a Heroku approved way?

Unfortunately, no. Each dyno gets its own ephemeral filesystem, with a fresh copy of the most recently deployed code.
This means two things:
1- Each of your web dynos is going to have its own, different, logs/production.rb which I think is not ideal.
2- Every time your dyno is restarted, as a result of deployment or daily cycling, you lose your logs/production.rb and start fresh. This I'm guessing would cause some logs to be lost.

> Might this approach cause our production server to go down?

Writing too much to your dyno will eventually cause your dyno to crash and follow [dyno crash restart policy](https://devcenter.heroku.com/articles/dynos#dyno-crash-restart-policy).
Unfortunately, we won't be able to give you any recommendations here since we're not familiar enough with datadog. but we know that writing your logs to a dyno is not recommended.


That means I have two choices:

1. Send logs via a drain to Datadog, no parsing of sidekiq logs, no matching to traces etc (see points others have raised above)
2. Send logs to the agent via a file, which could cause a dyno crash, lost logs, possibly duplicated logs

Neither of these solutions is adequate for us.

What am I missing here? How do I get proper logging in Datadog?

There is a third solution that I don't see mentioned anywhere...

`semantic_logger` can [send logs via UDP](https://www.rubydoc.info/gems/semantic_logger/SemanticLogger/Appender/Udp) or TCP and the Datadog Agent seems to be able to collect them via the same mechanisms.

Why is this not the approved way? Would it work?

**Comment by jeffblake** [comment]
I went with option #1 and implemented custom loggers for Rails and Sidekiq, in JSON format, and also added custom parsers in Datadog to parse out Postgres logs (and convert to metrics), etc.
Still not perfect but it's getting there

**Comment by adiberk** [comment]
This is still very much needed - do we have any updates on when this will go in?
1. As OP suggests the dyno vs source issue still persists for metric logs.  I was able to fix it by forcing dyno to be set to source.
2.  If I follow instructions and set HEROKU_APP_NAME so that host uses the APPNAME with the dyno name I am unable to correlate metrics and logs!  ---- BUT, if I set DD_HOSTNAME then I am able to do this.  However as mentioned in the comments, this skews data since a host can have more than 1 dyno...
